Client/Client.Core/Client.Core.Static/Client.cpp

```diff
@@ -657,4 +706,92 @@ void Client::onDirectMessageCreateAnswer(Utility::DirectMessageStatus directMess
         Base::Logger::LogLevel::WARNING
     );
 }
+
+void Client::onConnectionInfoAnswer(const Models::ConnectionInfo& connectionInfo)
+{
+    _connectionInfo = connectionInfo;
+
+    _connection->setConnectionVerifier(std::make_shared<Base::Verifiers::HashVerifier>());
+}
+
+void Client::onKeyAgreement(const Models::KeyAgreementInfo& serverKeyAgreementInfo)
+{
+    bool isSharedSecretCalculated = false;
+    CryptoPP::SecByteBlock sharedSecret;
+
+    if (!serverKeyAgreementInfo._publicKey.empty())
+    {
+        std::string publicServerKeyStr = serverKeyAgreementInfo._publicKey;
+        CryptoPP::SecByteBlock publicServerKey(reinterpret_cast<const CryptoPP::byte*>(publicServerKeyStr.data()), publicServerKeyStr.size());
+        sharedSecret = _connection->getKeyAgreement()->calculateSharedKey(publicServerKey);
+
+        isSharedSecretCalculated = !sharedSecret.empty();
+    }
+    else
+    {
+        Base::FileLogger::getInstance().log("Empty public server key. ", Base::LogLevel::WARNING);
+    }
+
+    if (isSharedSecretCalculated && !serverKeyAgreementInfo._publicKey.empty() &&
+        Base::SessionKeyHolder::Instance().setKey(std::move(sharedSecret), _connection->getUserID()) == Utility::GeneralCodes::SUCCESS)
+    {
+        _connection->setEncryption(std::make_shared<Base::Crypto::Symmetric::AES_GCM>());
+        _connection->setKeyConfirmator(std::make_shared<Base::KeyConfirmators::KeyConfirmation<> >());
+
+        _connection->getKeyAgreement().~shared_ptr();
+
+        Message messageToConfirmSharedKey;
+        messageToConfirmSharedKey.mHeader.mMessageType = Message::MessageType::KeyConfirmation;
+        messageToConfirmSharedKey.mBody = std::make_any<std::string>(_connection->getKeyConfirmator()->getVerificationUnit());
+
+        send(messageToConfirmSharedKey);
+    }
+    else
+    {
+        Base::SessionKeyHolder::Instance().Instance().removeKey(_connection->getUserID());
+
+        Base::FileLogger::getInstance().log(
+            std::string("Try to generate encryption key again. Current attempt: " + std::to_string(serverKeyAgreementInfo._attempt)),
+            Base::LogLevel::WARNING);
+
+        Message tryToGenerateSharedKeyAgain = constructKeyAgreementMessage(serverKeyAgreementInfo._attempt + 1);
```

Comment:
I guess you are using this class variable _attempt to stop when value overcomes limit, am I right?